### PR TITLE
Refactor i18n and add additional translation keys

### LINKED ISSUE #476: Additional items for internationalization
We are missing i18n keys for a number of UI items (and tooltip/title attributes). Here are items discovered to date (moved from #388):

- Next Hint
- Copy to Clipboard
- Solution
- R code

Please let me know if we've missed any other UI text. If possible, it would be helpful to include a screenshot.

### PR DESCRIPTION
**Pull request by gadenbuie**
I've refactored the internationalization scripts to remove the jquery-i18next dependency in favor of a hand-rolled function that achieves similar results, but lets us store some of the i18n parameters in `data-i18n-` HTML element attributes.

I also plan to refactor the `i18n_translations()` function to use a better format and preparation process that will be more robust for Unicode characters not in UTF-8.

The big change here is that localization will now:

- Use the key in the `data-i18n` attribute of the HTML element to look up the translation
- OR use the key in the `data-i18n-attr-<name>` attribute to look up the translation for the `<name>` attribute.
    - e.g. `data-i18n-attr-title = "button.runcodetitle"` uses the `button.runcodetitle` key to set the `title` attribute of the current element.
    - These elements must have a `data-i18n` attribute for internationalization to occur, if the innerHTML is not to be set, then set `data-i18n = ""`.
- Additional options to be passed to the `i18next.t()` method can be stored in the `data-i18n-opts` attribute. This is particularly useful for passing `{count: collection.length}` for pluralization.

I added an R function helper `i18n_span()` to help set the above attributes from the R side. For buttons created in JavaScript, these attributes are added as/where needed.

PR task list:
- [x] Update NEWS
- [x] Add tests (if possible)
- [x] Update documentation with `devtools::document()`

Fixes #476

### CODE REVIEW

File: data-raw/i18n_translations.yml [hunk 1]
```diff
@@ -0,0 +1,170 @@
+#### NOTE: This file is compiled for {learnr} by data-raw/i18n_translations.R ####
```
Comment (schloerke) [inline data-raw/i18n_translations.yml:1]:
```suggestion
#### NOTE: This file is the original data for data-raw/i18n_translations.R ####
```

File: inst/lib/tutorial/tutorial.js [hunk 2]
```diff
@@ -1190,7 +1200,10 @@ Tutorial.prototype.$addSolution = function(exercise, panel_heading, editor) {
     var hintIndex = 0;
 
     // create solution buttion
-    var button = addHintButton(caption);
+    var button = addHintButton(
+      caption,
```
Comment (schloerke) [inline inst/lib/tutorial/tutorial.js:1204]:
```suggestion
      isSolution ? 'Solution' : (hints.length > 1 ? 'Hints' : 'Hint'),
```

File: inst/lib/tutorial/tutorial.js [hunk 3]
```diff
@@ -1164,18 +1182,10 @@ Tutorial.prototype.$addSolution = function(exercise, panel_heading, editor) {
...
-    if (solution) {
-      caption = "Solution";
-    }
-    else {
-      if (hints.length > 1)
-        caption = "Hints";
-      else
-        caption = "Hint";
-    }
+    var caption = isSolution ? 'Solution' : (hints.length > 1 ? 'Hints' : 'Hint');
```
Comment (schloerke) [inline inst/lib/tutorial/tutorial.js:1188]: Let's remove this line in favor of inlining it below

File: inst/lib/i18n/tutorial-i18n-init.js [hunk 4]
```diff
@@ -64,17 +101,19 @@ $(document).on("shiny:sessioninitialized", function() {
...
     }
     i18next.changeLanguage(lang);
-    $(selector).localize(opts);
+    localize(selector, opts);
   }
 
   // localize question buttons when shown
   $(document).on('shiny:value', '.tutorial-question', function(ev) {
-    setTimeout(function() { $(ev.target).localize() }, 0);
+    setTimeout(function() {
```
Comment (schloerke) [inline inst/lib/i18n/tutorial-i18n-init.js:114]:
```suggestion
    // Allow for the DOM to be updated before updating
    setTimeout(function() {
```

File: inst/lib/i18n/tutorial-i18n-init.js [hunk 5]
```diff
@@ -45,15 +45,52 @@ $(document).on("shiny:sessioninitialized", function() {
...
+      if (this.dataset.i18nOpts) {
+        optsItem = Object.assign(optsItem, JSON.parse(this.dataset.i18nOpts));
+      }
+
+      // Translate the item iteslf
+      if (this.dataset.i18n) {
+        this.innerHTML = i18next.t(this.dataset.i18n, optsItem);
+      }
+
+      // Translate attribute values, getting keys from data-i18n-attr-<value>
```
Comment (schloerke) [inline inst/lib/i18n/tutorial-i18n-init.js:74]:
```suggestion
      // Translate attribute values, getting _all_ keys from data-i18n-attr-<VALUE>
```

File: inst/lib/i18n/tutorial-i18n-init.js [hunk 6]
```diff
@@ -45,15 +45,52 @@ $(document).on("shiny:sessioninitialized", function() {
...
+    });
+
+    if (!$els.length) {
+      // console.error('No elements found for localization with selector ' + selector);
+      return;
+    }
+
+    $els.each(function(idx) {
+      var optsItem = Object.assign({}, opts);
+
```
Comment (schloerke) [inline inst/lib/i18n/tutorial-i18n-init.js:63]:
```suggestion

      // Note: `this.dataset.i18nOpts` maps directly to the DOM element attribute `data-i18n-opts`
      //   And `this.dataset.i18nAttr` to element attribute `data-i18n-opts`
      // Link: https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes
      // Reference:
      // > To get a data attribute through the dataset object, get the property by the part of the attribute name after data- (note that dashes are converted to camelCase).

```

**Review by schloerke — APPROVED** [review]
LGTM given small code shuffle / notes

**Comment by gadenbuie** [comment]
> LGTM given small code shuffle / notes

Thanks @schloerke! I've updated the comments and shuffled the code as requested 😄

**Pull request merged**